backend/wmg/api/v1.py

```diff
@@ -1,58 +1,141 @@
+# TODO: This needs a test, but since we are likely to replace it (see above TODO), it is currently not tested
+def fetch_datasets_metadata(dataset_ids: List[str]) -> List[Dict]:
+    # We return a DTO because the db entity can't access its attributes after the db session ends,
+    # and we want to keep session management out of the calling method
+
+    with db_session_manager() as session:
+
+        def get_dataset(dataset_id_):
+            dataset = Dataset.get(session, dataset_id_)
+            if dataset is None:
+                # Handle possible missing dataset due to db state evolving past wmg snapshot
+                return dict(id=dataset_id_, label="", collection_id="", collection_label="")
+            return dict(
+                id=dataset.id,
+                label=dataset.name,
+                collection_id=dataset.collection.id,
+                collection_label=dataset.collection.name,
+            )
+
+        return [get_dataset(dataset_id) for dataset_id in dataset_ids]
+
+
+def build_filter_dims_values(all_filter_dims_values):
+    response_filter_dims_values = dict(
+        datasets=fetch_datasets_metadata(all_filter_dims_values["dataset_id"]),
+        disease_terms=build_ontology_term_id_label_mapping(all_filter_dims_values["disease_ontology_term_id"]),
+        sex_terms=build_ontology_term_id_label_mapping(all_filter_dims_values["sex_ontology_term_id"]),
+        development_stage_terms=build_ontology_term_id_label_mapping(
+            all_filter_dims_values["development_stage_ontology_term_id"]
+        ),
+        ethnicity_terms=build_ontology_term_id_label_mapping(all_filter_dims_values["ethnicity_ontology_term_id"]),
+        # excluded per product requirements, but keeping in, commented-out, to reduce future head-scratching
+        # assay_ontology_terms=build_ontology_term_id_label_mapping(all_filter_dims_values["assay_ontology_term_id"]),
+    )
+    return response_filter_dims_values
+
+
+def build_expression_summary(query_result: DataFrame) -> dict:
+    # Create nested dicts with gene_ontology_term_id, tissue_ontology_term_id keys, respectively
+    structured_result = defaultdict(lambda: defaultdict(list))
+    for group_by_key, cell_type_stats in query_result.to_dict("index").items():
+        gene_ontology_term_id, tissue_ontology_term_id, cell_type_ontology_term_id = [s for s in group_by_key]
+        structured_result[gene_ontology_term_id][tissue_ontology_term_id].append(
+            dict(
+                id=cell_type_ontology_term_id,
+                n=cell_type_stats["nnz"],
+                me=cell_type_stats["sum"] / cell_type_stats["n_cells"],
+                # TODO
+                pc=0.0,
+                # TODO
+                tpc=0.0,
+            )
+        )
+    return structured_result
+
+
+def extract_filter_dims_values(query_result: DataFrame) -> dict:
+    """
+    Return unique values for each dimension in the specified query result
+    """
+    return {
```

Comment:
nit I think this might be a little more readable if you decomposed it to something like this (assuming I'm groking the original function correctly)

some_dict = {}
for col in query_result.columns:
    some_keys = query_result.groupby(col).groups.keys()
    some_dict[col] = set(some_keys) & set(cube_non_indexed_dims)
    
return some_dict

Reply:
Was trying to remain Pythonic in use of the dictionary-producing comprehension. Let me take a stab at reworking for clarity.